Replace calls to contiguous with contiguous(suggested memory format)

kimishpatel opened the pull request:
Summary:
Wherever applicable it would be better to call contiguous with appropriate
memory format.
Plus output should be allocated with the same memory format as input when
applicable. Otherwise convert to that format upon returning.
This helps with some perf where otherwise calls to contiguous may involve
allocation and memcpy.

Test Plan: quantization tests

Differential Revision: D21559301

vkuzo commented:
> Plus output should be allocated with the same memory format as input when
applicable. Otherwise convert to that format upon returning.

thoughts on whether this applies well to ops which have kernels which are faster in one of the formats?  I.e. if op 1 outputs NCHW, op 2 needs NHWC for a fast kernel, and op 3 needs NHWC for a fast kernel, it would be faster for op 2 to output the format it ran the calculation in.  Note: this is a theoretical question, I didn't look up how common this is.

vkuzo commented:
logic lgtm, after fixing the quantized test failures should be good to go!

kimishpatel commented:
> > Plus output should be allocated with the same memory format as input when
> > applicable. Otherwise convert to that format upon returning.
> 
> thoughts on whether this applies well to ops which have kernels which are faster in one of the formats? I.e. if op 1 outputs NCHW, op 2 needs NHWC for a fast kernel, and op 3 needs NHWC for a fast kernel, it would be faster for op 2 to output the format it ran the calculation in. Note: this is a theoretical question, I didn't look up how common this is.

I agree that it should be converted in the format for which the fast kernel exist. We dont seem to have a unified policy around this though and this is a hard question to answer. You have to tradeoff the cost of memory allocation copying against how much fast kernel buys us.
Having said that there are places I see where we opt for specialized kernels if one exist for the memory format of the input tensor, but never explicitly convert. Only exception I saw was group norm, where it made sense.

With respect to returning the output tensor in the same format as input: This is the op level agreement established from what I understand. @gchanan can say more about it. However right now this is not conformed to in many places, including quantized kernels.

kimishpatel commented:
@supriyar @vkuzo, for some reason relu does not like `.contiguous(self.suggest_input_format()`. I tried to understand why but could not quite figure out. It is strange that it forces you to do `.contiguous()`. You guys may have a better historical context about this.

vkuzo commented:
> @supriyar @vkuzo, for some reason relu does not like `.contiguous(self.suggest_input_format()`. I tried to understand why but could not quite figure out. It is strange that it forces you to do `.contiguous()`. You guys may have a better historical context about this.

is there a stack trace of what breaks for ReLU?

kimishpatel commented:
> > @supriyar @vkuzo, for some reason relu does not like `.contiguous(self.suggest_input_format()`. I tried to understand why but could not quite figure out. It is strange that it forces you to do `.contiguous()`. You guys may have a better historical context about this.
> 
> is there a stack trace of what breaks for ReLU?

Stack trace? The test fails. It produces wrong output. I think it should be locally reproducible on master.

supriyar commented on aten/src/ATen/native/quantized/cpu/qrelu.cpp:
```diff
@@ -24,7 +24,7 @@ Tensor qnnpack_relu(Tensor input) {
       input.ndimension() > 0, "qnnpack_relu(): Got empty input tensor");
 
-  Tensor input_contig = input.contiguous();
+  Tensor input_contig = input.contiguous(input.suggest_memory_format());
```
@kimishpatel are you saying this line breaks tests?

kimishpatel replied:
Yes. I had to back out the change because it was breaking this test: `quantization.test_quantize.TestModelNumerics. test_float_quant_compare_per_tensor`. If I removed the second relu in the module (https://github.com/pytorch/pytorch/blob/87b198d309e00f52e098a533c22b087b36fc00e3/torch/testing/_internal/common_quantization.py#L773) for this test it worked. First one gets fused.

vkuzo replied:
I didn't have time to finish the investigation yet, but hope this helps:
* adding some debug code to narrow down the problem: https://gist.github.com/vkuzo/9e6fe1e4ef0a4754aef3026e97244240 .
* with the code above
`
# correct layer output
[0.8699, 0.8249, 0.8849, 0.8399, 0.7949, 0.7349, 0.7349, 0.7199, 0.8399, 0.8849, 0.7199, 0.7199]
# incorrect layer output
[0.8699, 0.7949, 0.8399, 0.8249, 0.7349, 0.8849, 0.8849, 0.7349, 0.7199, 0.8399, 0.7199, 0.7199]
# we see that incorrect layer output is the NHWC version of the correct NCHW layer output
`
so, something seems funky with the memory format of the output

vkuzo reviewed (approved):
lg!

kimishpatel commented:
OK so the conclusion from the relu debug is:
`input.options()` return (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/templates/TensorMethods.h#L50) do not contain memory format information. Thus this is always expected to be separately specified. I am updating this PR to account for that in relu.

kimishpatel committed:
Replace calls to contiguous with contiguous(suggested memory format) (#38433)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/38433

Wherever applicable it would be better to call contiguous with appropriate
memory format.
Plus output should be allocated with the same memory format as input when
applicable. Otherwise convert to that format upon returning.
This helps with some perf where otherwise calls to contiguous may involve
allocation and memcpy.

Test Plan: quantization tests

Reviewed By: vkuzo

Differential Revision: D21559301

fbshipit-source-id: 1d72fc8d896ccca9addb0b0ad55c492561049e87
